client/me/purchases/pre-cancellation-dialog/index.tsx

```diff
@@ -0,0 +1,226 @@
+import { Dialog, Gridicon } from '@automattic/components';
+import cx from 'classnames';
+import { useTranslate } from 'i18n-calypso';
+import FormattedHeader from 'calypso/components/formatted-header';
+import { SiteExcerptData } from 'calypso/data/sites/site-excerpt-types';
+import {
+	hasAmountAvailableToRefund,
+	isRefundable,
+	maybeWithinRefundPeriod,
+} from 'calypso/lib/purchases';
+import { CALYPSO_CONTACT } from 'calypso/lib/url/support';
+import getPlanCancellationFeatures from './get-plan-cancellation-features';
+import type { Purchase } from 'calypso/lib/purchases/types';
+import './style.scss';
+
+interface PreCancellationDialogProps {
+	closeDialog: () => void;
+	removePlan: () => void;
+	isDialogVisible: boolean;
+	isRemoving: boolean;
+	site: SiteExcerptData;
+	purchase: Purchase;
+	hasDomain: boolean;
+	primaryDomain: string;
+	wpcomURL: string;
+}
+
+export const PreCancellationDialog = ( {
+	closeDialog,
+	removePlan,
+	isDialogVisible,
+	site,
+	purchase,
+	hasDomain,
+	primaryDomain,
+	wpcomURL,
+}: PreCancellationDialogProps ) => {
+	const translate = useTranslate();
+
+	/**
+	 * Instantiate site's plan variables.
+	 */
+	const productSlug = site.plan?.product_slug;
+	const planLabel = site.plan?.product_name_short;
+	const subTitle = translate( 'Subtitle' );
+	const isPurchaseRefundable = isRefundable( purchase );
+	const isPurchaseAutoRenewing = purchase.isAutoRenewEnabled;
+
+	/**
+	 * Click events, buttons tracking and action.
+	 */
+	const clickCancelPlan = () => {
+		removePlan();
+	};
+
+	const clickCloseDialog = () => {
+		closeDialog();
+	};
+
+	/**
+	 * Domain redirect copy
+	 */
+	const domainFeature =
+		hasDomain && primaryDomain && wpcomURL ? (
+			<p>Warning: domain traffic will be redirected</p>
+		) : null;
+
+	/**
+	 * Dialog buttons
+	 */
+	const buttons = [
+		{
+			action: 'cancel',
+			label: translate( 'Cancel my plan' ),
+			onClick: clickCancelPlan,
+		},
+		{
+			action: 'close',
+			label: translate( 'Keep my plan' ),
+			onClick: clickCloseDialog,
+			isPrimary: true,
+		},
+	];
+
+	/**
+	 * Render dialog bottom text:
+	 * - Link to contact support.
+	 * - If refundable, the amount to be refunded.
+	 */
+	const RenderFooterText = () => {
```

Comment:
It's probably fine to have these components defined inline inside another component, but since that will cause the component definitions to change on every render, it might not be as performant as if this was a stand-alone component. Would it be difficult to move this and `FeaturesList` out of this function to the top-level of this file? Not a blocker if not.

Reply:
@sirbrillig thanks for this comment, I refactored the component as requested